sdk/identity/identity/src/credentials/interactiveBrowserCredential.browser.ts

```diff
@@ -55,46 +80,95 @@ export class InteractiveBrowserCredential implements TokenCredential {
     const knownAuthorities =
       options.tenantId === "adfs" ? (options.authorityHost ? [options.authorityHost] : []) : [];
 
+    this.correlationId = options.correlationId;
+
     this.msalConfig = {
       auth: {
         clientId: options.clientId!, // we just initialized it above
         authority: `${options.authorityHost}/${options.tenantId}`,
-        knownAuthorities,
-        ...(options.redirectUri && { redirectUri: options.redirectUri }),
-        ...(options.postLogoutRedirectUri && { redirectUri: options.postLogoutRedirectUri })
+        knownAuthorities
       },
       cache: {
-        cacheLocation: "localStorage",
-        storeAuthStateInCookie: true
+        cacheLocation: "sessionStorage",
+        storeAuthStateInCookie: true // Set to true to improve the experience on IE11 and Edge.
+      },
+      system: {
+        loggerOptions: {
+          loggerCallback: (level, message, containsPii) => {
+            if (containsPii) {
+              return;
+            }
+            switch (level) {
+              case msalBrowser.LogLevel.Error:
+                logger.info(`MSAL Browser V2 error: ${message}`);
+                return;
+              case msalBrowser.LogLevel.Info:
+                logger.info(`MSAL Browser V2 info message: ${message}`);
+                return;
+              case msalBrowser.LogLevel.Verbose:
+                logger.info(`MSAL Browser V2 verbose message: ${message}`);
+                return;
+              case msalBrowser.LogLevel.Warning:
+                logger.info(`MSAL Browser V2 warning: ${message}`);
+                return;
+            }
+          }
+        }
       }
     };
 
-    this.msalObject = new msal.UserAgentApplication(this.msalConfig);
+    this.msalConfig.auth.redirectUri =
+      typeof options.redirectUri === "function" ? options.redirectUri() : options.redirectUri;
+
+    this.msalConfig.auth.postLogoutRedirectUri =
+      typeof options.postLogoutRedirectUri === "function"
+        ? options.postLogoutRedirectUri()
+        : options.postLogoutRedirectUri;
+
+    this.msalObject = new msalBrowser.PublicClientApplication(this.msalConfig);
+
+    this.handleRedirectResponse();
```

Comment:
**Important:**  
After redirect, the first time the user initializes any `InteractiveBrowserCredential` is when we'll be able to authenticate based on the redirect response. To finish the authentication flow, we need the fully formed configuration object, which we generate based on the parameters to this credential. Is it safe to assume that the first initialization of `InteractiveBrowserCredential` can be used to finish the MSAL redirect authentication flow?

An alternative I can think of is to store each set of new `InteractiveBrowserCredential` parameters in something like `localStorage`, then to cache the promise on page load based on the latest set of parameters, then on constructor initialization to match the promise based on parameters again.

My thought process is that I believe we can get away with asking the user to instantiate the `InteractiveBrowserCredential` on page load for the redirect style. Is this a reasonable assumption?

@jonathandturner , @sameerag

Reply:
I believe `msal-browser` already addresses this. @pkanher617  to confirm.

Reply:
The main issue is that if I have an app that has two instances of this credential, without this cache, the first one that is able to handle the redirect will get authenticated, the second one won't 🤔 I could be doing something wrong though.

Also, a separate problem perhaps:
Consider an application trying to authenticate two different accounts, how would one identify each individual redirect response? One would happen first, of course, then the other one would need to happen, but each time the page loads we lose track of the other credential. Does that make sense? I wonder if I'm doing something wrong.

Reply:
Ok so we arrived at the following:

- We'll only allow one active account per client ID and tenant ID pair.
- If two or more clients try to authenticate with the same client ID and tenant ID, if the first succeeds authenticating, the others will be able to retrieve this same account, so there will be no further authentication attempts, and our limitation above will be respected: only one account per client ID and tenant ID.
- If the user happens to use MSAL-browser by their own account, and the storage is polluted, we might be able to get into two accounts. In that case, we will throw with an error, asking them to clear their browser cache.

I'll be asking @schaabs about this approach, but in the mean time, this is the update I'll be making to this PR.